server.go

```diff
@@ -4173,13 +4327,297 @@ func (s *server) SubscribeCustomMessages() (*subscribe.Client, error) {
+	pendingChan *channeldb.OpenChannel, remotePub string) error {
+
+	// Call NewPendingOpenChan to update the server's maps for this peer.
+	if err := s.NewPendingOpenChan(remotePub); err != nil {
+		return err
+	}
+
+	// Notify subscribers about this event.
+	s.channelNotifier.NotifyPendingOpenChannelEvent(op, pendingChan)
+
+	return nil
+}
+
+// notifyFundingTimeoutPeerEvent updates the server's maps and then calls the
+// channelNotifier's NotifyFundingTimeout.
+func (s *server) notifyFundingTimeoutPeerEvent(op wire.OutPoint,
+	remotePub string) error {
+
+	// Call NewPendingCloseChan to potentially demote the peer.
+	if err := s.NewPendingCloseChan(remotePub); err != nil {
+		return err
+	}
+
+	// Notify subscribers about this event.
+	s.channelNotifier.NotifyFundingTimeout(op)
+
+	return nil
+}
+
+// NewPendingOpenChan is called after the pending-open channel has been
+// committed to the database. This may transition a restricted-access peer to a
+// temporary-access peer.
+func (s *server) NewPendingOpenChan(remoteHex string) error {
+	s.banScoreMtx.Lock()
+	defer s.banScoreMtx.Unlock()
+
+	// Fetch the peer's access status from peerScores.
+	status, found := s.peerScores[remoteHex]
+	if !found {
+		// If we didn't find the peer, we'll return an error.
+		return ErrNoPeerScore
+	}
+
+	switch status.state {
+	case protected:
+		// If this peer's access status is protected, we don't need to
+		// do anything.
+		return nil
+
+	case temporary:
+		// If this peer's access status is temporary, we'll need to
+		// update the pendingChannelPeers map. The peer's access status
+		// will stay temporary.
+		numPending, pendingFound := s.pendingChannelPeers[remoteHex]
+		if !pendingFound {
+			// Error if we did not find any info in
+			// pendingChannelPeers.
+			return ErrNoPendingPeerInfo
+		}
+
+		// Increment the numPending amount.
+		s.pendingChannelPeers[remoteHex] = numPending + 1
+
+	case restricted:
+		// If the peer's access status is restricted, then we can
+		// transition it to a temporary-access peer. We'll need to
+		// update numRestricted and also peerScores. We'll also need to
+		// update pendingChannelPeers.
+		s.pendingChannelPeers[remoteHex] = 1
+
+		// A restricted-access slot has opened up.
+		s.numRestricted -= 1
+
+		s.peerScores[remoteHex] = peerSlotStatus{state: temporary}
+
+	default:
+		// This should not be possible.
+		return fmt.Errorf("invalid peer access status")
+	}
+
+	return nil
+}
+
+// NewPendingCloseChan is called when a pending-open channel prematurely closes
+// before the funding transaction has confirmed. This potentially demotes a
+// temporary-access peer to a restricted-access peer. If no restricted-access
+// slots are available, the peer will be disconnected.
+func (s *server) NewPendingCloseChan(remoteHex string) error {
+	// Ensure that we don't lock banScoreMtx and _then_ lock the server
+	// mutex. This could lead to deadlock.
+	s.banScoreMtx.Lock()
```

Comment:
Why not `defer s.banScoreMtx.Unlock()`? We unlock on every return path.

Reply:
See comment: 
```
// The banScoreMtx cannot be held when the
// below DisconnectPeer call occurs.
```
which is annoying because it means we can't `defer`